### DISCUSSION TITLE
Woodcutting plugin: Tree depletion timer

### DISCUSSION THREAD

**Discussion by Hjaldr** [runelite/runelite#16894] [Ideas]
There is a respawn timer on tree stumps, but not a 'falling' timer when a tree is starting to get chopped down. My suggestion is to add a similar optional timer that shows how long a tree has left before it is chopped down/depleted.

Relevant to consider is that the timer starts going up when nobody is chopping, but a plugin hub plugin called Tree Count already shows how many people are chopping which trees, so I reckon that it should be possible to track client-side which trees within render distance are counting down their depletion timer.

It would be as simple as displaying a timer with a countdown of the depletion timer for each tree, which only shows if the player is woodcutting (or perhaps even only if the player is on that particular tree). When nobody is chopping a tree, the timer could go back up or disappear.

**Comment by brads93** [comment]
I've added the wiki and a screenshot detailing the various despawn timers. Personally, I'm either going to stay at redwoods for afk or chop yews for forestry events, but many others have expressed an interest for a de-spawn timer since forestry released. Thanks for posting.
https://oldschool.runescape.wiki/w/Woodcutting
![image](https://github.com/runelite/runelite/assets/1676751/8abbf30c-a62f-4c74-9bde-690fe0579d01)

**Comment by CreativeTechGuy** [comment]
Can someone from the RuneLite team confirm if this feature request would be accepted to the core Woodcutting plugin? If so, I'd make it. But I want to make sure that it would be approved first. @Adam- @Nightfirecat ?

**Reply by DominickCobb-rs** [reply]
I'm fairly certain Adam has expressed he has no interest in this being added to core. See [this thread](https://discord.com/channels/301497432909414422/419891709883973642/1126699638859243531) in the discord

**Reply by CreativeTechGuy** [reply]
Reading that it just looks like we need to figure out the regen rate of the tree and he'd be happy to include it. The only concern he had was that it might be inaccurate. It seems like if we tested it we could get accurate numbers.

It seems like it would be pretty easy to test:
1. Start chopping a tree for 15 seconds
2. Stop chopping the tree for 10 seconds
3. Start chopping the tree again and check how long it takes to fall down.

Since we know exactly how long it takes to fall if chopping uninterrupted, we can then find the difference of how much regenerated during that gap. Then we'd just do that with several other intervals and tree types to confirm the hypothesis.

**Reply by Hjaldr** [reply]
I would gladly do some testing - If you can give me some directions as to how you'd want the data to be formatted I can get you data for all trees.

**Reply by CreativeTechGuy** [reply]
@Hjaldr I've actually been working on a plugin for this since my last message and it's basically complete at this point. I've done a bunch of investigation and it turns out the system isn't super precise. I'll list what my current hypothesis is for how it works, it'd be amazing if you could confirm/deny my hypothesis. It'd help make the plugin more accurate!

* All trees are the same. I cannot see any difference in the mechanics per tree, although the despawn timers for each tree are different.
* The timer only starts once you collect the first log from the tree.
  * I am unsure if this condition resets if all people stop chopping a tree, let it partially/fully regenerate, if the timer will only start counting down again once you get another log. That's super hard to test.
  * One way you could test this is by stopping chopping a tree after you get every log and immediately restarting (assuming you are chopping alone). If the timer only ticks down after chopping the first log and that tracker resets whenever no one is chopping, then this should make the tree last significantly longer than chopping it continuously. And if so, then we'd need to figure out if the tree timer is paused or regenerating at that time.
* The rate for despawn and regen seems to be the same at 1 tick up or down per game tick (0.6s).
* The tree will only fall if the despawn timer is zero *and* someone gets a log from the tree. It'll stay standing past the timer if no one collects a log.
  * I think that if the timer is zero and everyone stops chopping before getting another log, it'll still be able to regenerate at this point. I am not sure if the tree can go negative during this process though or if it bottoms out at zero. Also super hard to test.
* I think using a dragon axe special attack while chopping the tree will continue counting down the timer despite you not doing the chopping animation during the special. That is just a guess though since you automatically resume chopping after the special animation is done. Possibly the tree does regenerate for the 6 ticks that the animation is playing.
* I assume the trees regenerate even when no one is around. It'd be interesting to know if there's no players in the area and the tree is partially chopped, and then everyone leaves if the areas unloads, resets, or continues regenerating at 1 per tick like normal. This would probably be easiest to check in a dimension which has trees but very few players since you could guarantee that the entire dimension would be unloaded if you leave.
* I also have a hunch that the timers aren't exact, especially when chopping in mass. When chopping alone the above info looks super precise, but when you are in a forestry mass, it seems like trees can sometimes fall earlier than what the timer would suggest. It seems random though as sometimes it's perfectly on time, sometimes it's longer, and sometimes it's shorter. But I wonder if Jagex has some bug with lots of players.
* I've been assuming everything is happening on the tick system. But it's also possible that when Jagex said that trees despawn after a certain number of seconds, that they are actually tracking the seconds for real ignoring ticks. That might explain some of the inconsistencies since ticks aren't always a consistent 0.6s.

I know that's a lot, but I just wanted to give you a headstart on everything I've already researched. If you want to help, being able to test any of those things would be super awesome. But honestly I'm not sure a good way to test them reliably.

**Reply by DominickCobb-rs** [reply]
> The timer only starts once you collect the first log from the tree.

I believe this is the reason why it's not being sought for core. It's unreliable and nearly impossible to do correctly 100% of the time.

**Reply by CreativeTechGuy** [reply]
> I believe this is the reason why it's not being sought for core. It's unreliable and nearly impossible to do correctly 100% of the time.

I understand that is their philosophy, but I don't agree with it. It seems like a mistake to not seek out more ambitious plugins simply because they aren't perfect. A timer indicating how long you'll be able to AFK for is incredibly useful even if it is off by 5-10% sometimes. It's worse for everyone to have too many 3rd party plugins as that just creates more opportunities for plugins to be broken or abandoned, or have bugs, inconsistencies, or incompatibilities with each other.

The [Guardians of the Rift Helper](https://runelite.net/plugin-hub/show/guardians-of-the-rift-helper) is the prime example of this. It's a plugin that has been mostly abandoned and has countless bugs which will probably never be fixed. It's a real shame given that it's the 3rd most popular plugin on Plugin Hub. Something that would have never happened if it was in the core (or at the very least a community-owned repo). It would be very interesting if all plugin hub plugins could exist in a monorepo so that ownership could easily change. A model like how [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) works would be great. Anyway I digress.

I'm already making this as a 3rd party plugin because I know that's the only way it'll be accepted. But I wish that the RuneLite team would be more open to plugins that are mostly correct as they are still more helpful than if they didn't exist at all.

**Reply by brads93** [reply]
Love that you are dedicating some time to this. If there is any way I can help test, let me know. I have access to all trees and wanted a tree despawn timer while I was training forestry. I love runelite and anything the community can do to make our lives easier, and this is a specific plugin I've been looking forward to. Hit me up if you need help testing. Discord: ong_frfr_nocap

**Reply by CreativeTechGuy** [reply]
@brads93 Hey thanks for the offer! I posted a long comment [a bit higher in this thread](https://github.com/runelite/runelite/discussions/16894#discussioncomment-6908081) with details on what I currently have researched. If you'd like to help confirm/deny any of that with your own findings and post it back here, that'd be great.
